### ISSUE TITLE
Reading file list from stdin chokes on large files

### ISSUE THREAD

**Issue by janbernloehr** [microsoft/pyright#9583]
**Describe the bug**
When passing a large list of files names to pyright to read from stdin, we get an `Invalid file list specified by stdin input` error because [we read from stdin with `fs.readFileSync(process.stdin.fd, 'utf-8')`](https://github.com/microsoft/pyright/blob/9041f88758e94a5419a9b07e9f73b37eecc6dc06/packages/pyright-internal/src/pyright.ts#L258) but for a large input, the data arrives in chunks.

**Version**
```sh
$ pyright --version
1.1.390
```

**How to reproduce**
```sh
$ cat filenames.txt | pyright -
Invalid file list specified by stdin input.
```

you can use this script to generate a large number of random file names to trigger the bug

```python
import string, random

count = 10000
min_length = 100
output_file = "filenames.txt"
characters = string.ascii_letters + string.digits
random_length = min_length - 2

with open(output_file, 'w') as f:
    for _ in range(count):
        fname = ''.join(random.choices(characters, k=random_length)) + '.py\n'
        f.write(fname)

```

**Comment by erictraut** [comment]
Thanks for the bug report. 

You're using this feature in a manner that wasn't originally intended, but this failure is one that I wouldn't expect. It appears to be a limitation of the `readFileSync` call in node when used with `stdin`. I'll see if there's a straightforward fix with minimal regression risk. I think it's safe to say that you (among the millions of pyright and pylance users) are the only one hitting this issue, so if this turns out to be difficult or risky to fix, I may choose not to address it and as you to implement your own workaround.

I'm curious to learn more about what you're doing here. Why are you using this feature to create a manual list of tens of thousands of files to send to the pyright command line rather than using a config file, as is typical?

**Comment by erictraut** [comment]
This turned out to be easy to fix. It will be addressed in the next release.

**Comment by janbernloehr** [comment]
On why we are doing this: We have a mono repo with ~10k files. We run pyright as part of our CI on every PR. For the whole repo it takes about 8min but we use static code analysis to build an import graph to run it only on the change set plus all files which (recursively) import any changed file. This dramatically speeds things up. However, the change sets can become very big and that’s where we saw this issue.

Note: we initially used to pass just the file names via command line however we also hit the max arglenth limits with out repo, that’s why we looked into alternative ways to pass a large list of files.

**Comment by erictraut** [comment]
Interesting approach. Thanks for the details.

Eight minutes is a long time, even for 10K files. I wonder if there's a particular file (or a small subset of files) that account for the majority of this time. If you're interested in investigating that possibility, you could use `--stats --verbose` when running pyright. It will then report the time it takes per file sorted in decreasing order.

**Comment by janbernloehr** [comment]
When running on 1.1.390 on my repo I get 182x "Long operation: analyzing" and 176x "Long operation: checking" the longest being ~9s and the shorted ~2s

**Comment by erictraut** [comment]
That message indicates that it took more than 2000ms to analyze a file. In most source bases, this is rare. It can happen in cases that involve deeply-nested loops, large unions, complex code flows, or other programming constructs that can create "combinatoric explosions" for a static analyzer. Optimizations are sometimes possible if I'm able to reproduce the cause of the issue locally. If this is a public code repo, I could take a look. If it's private, then I'd need you to send me a simplified example that repros the issue if you'd like me to investigate further.

Another option that you might be interested in... A few months ago I added multi-core support to the command-line version of pyright. You can specify the "--threads" option to enable this. Pyright will then use multiple cores on your CI server to parallelize its analysis. For large code bases, this can typically cut the analysis time by 50% to 75%.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.391.
